### ISSUE TITLE
Extra completions don't work

### ISSUE THREAD

**Issue by pdrastil** [romkatv/zsh4humans#97]
Hi,
I was testing your v5 branch and it seems that extra completions provided by CLI utilities in my .zshrc are not loaded. Can you take a look if I'm doing something wrong or if this is broken?

```sh
z4h init || return
 
source <(oc completion zsh)
```

**Comment by pdrastil** [comment]
Link for CLI tool: https://mirror.openshift.com/pub/openshift-v4/clients/oc/4.5/linux/oc.tar.gz

**Comment by romkatv** [comment]
Thanks for the report. Fixed.

To pick up the fix you need to do the following:

1. Remove `source <(oc completion zsh)` from `~/.zshrc`.
2. Run `z4h update`.

**Comment by pdrastil** [comment]
Thanks, but I believe this is more generic issue. I have another non-public enterprise CLI tool that was build by my current customer and provides completions in the same way  `source <(enterprise-tool completion zsh)`. This one also doesn't work compared to v4. Would there be some better way how to hook this type of tools into your project?

**Comment by pdrastil** [comment]
I've also noticed that extra completions provided by `fpath=(~/.dotfiles/completion $fpath)` are also not available.

**Comment by romkatv** [comment]
> Thanks, but I believe this is more generic issue. I have another non-public enterprise CLI tool that was build by my current customer and provides completions in the same way  `source <(enterprise-tool completion zsh)`. This one also doesn't work compared to v4. Would there be some better way how to hook this type of tools into your project?

Maybe. The downside of making `source <(enterprise-tool completion zsh)` work is that I will not find out about public tools of this sort. If `source <(oc completion zsh)` worked, you wouldn't have filed this issue. And that would mean that zsh startup would be 150ms slower for you and for all `oc` users out there.

> I've also noticed that extra completions provided by `fpath=(~/.dotfiles/completion $fpath)` are also not available.

Please provide complete instructions how to reproduce this.

**Comment by pdrastil** [comment]
Would it be possibility to add zstyle option to disable this behaviour in v5? I'm completely fine with slower shell if that means I can fuzzy find and work with over 3000+ production clusters with my enterprise tool.

Regarding the `fpath` - just add extra folder in your default ~/.zshrc

Content of my folder typically contains extra completions for docker and docker-compose downloaded from:
-https://raw.githubusercontent.com/docker/docker-ce/master/contrib/completion/zsh/_docker
- https://raw.githubusercontent.com/docker/compose/master/contrib/completion/zsh/_docker-compose

Note that these scripts are being updated by wrapper script from time to time

**Comment by romkatv** [comment]
> Would it be possibility to add zstyle option to disable this behaviour in v5?

It's probably possible. It's not something I'm planning to do in the near future.

> Regarding the `fpath` - just add extra folder in your default ~/.zshrc

Please provide more specific instructions. Something like this:

1. Install Ubuntu 20.04.
2. Run the following commands:
   ```zsh
   sudo apt update
   sudo apt install -y curl zsh
   ```
3. Create `~/.zshenv` with the following content:
   ```text
   FIXME
   ```
4. Create `~/.zshrc` with the following content:
   ```text
   FIXME
   ```
5. Run `zsh`.
6. Type `blah -` and press <kbd>Tab</kbd>.
7. Etc.

Please verify that the issue reproduces if you follow your own instructions.

**Comment by pdrastil** [comment]
One more idea regarding the completion from CLI tools - allow users to define extra tools via zstyle that will be appended to well-known tools if the logic behind the completion is the same.

Reproduction:
1) OS: macOS Catalina / Big Sur
2) Create `.zshrc` with following content
```sh
# Personal Zsh configuration file. It is strongly recommended to keep all
# shell customization and configuration (including exported environment
# variables such as PATH) in this file or in files source by it.
#
# Documentation: https://github.com/romkatv/zsh4humans/blob/v5/README.md.

# Periodic auto-update on Zsh startup: 'ask' or 'no'.
zstyle ':z4h:'                auto-update      'ask'
# Ask whether to auto-update this often; has no effect if auto-update is 'no'.
zstyle ':z4h:'                auto-update-days '28'
# Automaticaly wrap TTY with a transparent tmux ('integrated'), or start a
# full-fledged tmux ('system'), or disable features that require tmux ('no').
zstyle ':z4h:'                start-tmux       'integrated'
# Move prompt to the bottom when zsh starts up so that it's always in the
# same position. Has no effect if start-tmux is 'no'.
zstyle ':z4h:'                prompt-at-bottom 'yes'
# Keyboard type: 'mac' or 'pc'.
zstyle ':z4h:bindkey'         keyboard 'mac'
# Right-arrow key accepts one character ('partial-accept') from
# command autosuggestions or the whole thing ('accept')?
zstyle ':z4h:autosuggestions' forward-char 'accept'
# Enable ('yes') or disable ('no') automatic teleportation of z4h over
# ssh when connecting to these hosts.
# example: zstyle ':z4h:ssh:example-hostname1'   enable 'yes'
# The default value if none of the overrides above match the hostname.
zstyle ':z4h:ssh:*'           enable 'no'
# Send these files over to the remote host when connecting over ssh to the
# enabled hosts.
zstyle ':z4h:ssh:*'           send-extra-files '~/.nanorc' '~/.env.zsh'

# Install or update core components (fzf, zsh-autosuggestions, etc.) and
# initialize Zsh. After this point console I/O is unavailable until Zsh
# is fully initialized. Everything that requires user interaction or can
# perform network I/O must be done above. Everything else is best done below.
z4h init || return

# ---[ Environment ]-------------------------------------------
export LANG=en_US.UTF-8
export DOTFILES="$HOME/.dotfiles"
export GPG_TTY=$TTY
export GOPATH=$(go env GOPATH)
export FZF_DEFAULT_COMMAND="fd --type f"

# Default editor for local and remove environments
if [[ -n "$SSH_CONNECTION" ]]; then
  # on server
  if command -v vim >/dev/null 2>&1; then
    export EDITOR='vim'
  else
    export EDITOR='vi'
  fi
else
  export EDITOR='nvim'
fi

# Extend PATH.
path=(~/bin /usr/local/sbin $path)
fpath=(~/.dotfiles/completion $fpath)

# Source additional local files if they exist.
z4h source ~/.env.zsh

# Define key bindings.
z4h bindkey undo           Ctrl+/       # undo the last command line change
z4h bindkey redo           Alt+/        # redo the last undone command line change
z4h bindkey z4h-cd-back    Shift+Left   # cd into the previous directory
z4h bindkey z4h-cd-forward Shift+Right  # cd into the next directory
z4h bindkey z4h-cd-up      Shift+Up     # cd into the parent directory
z4h bindkey z4h-cd-down    Shift+Down   # cd into a child directory

# Autoload functions.
autoload -Uz zmv

# Define functions and completions.
function md() { [[ $# == 1 ]] && mkdir -p -- "$1" && cd -- "$1" }
compdef _directories md

# ---[ Aliases ]-------------------------------------------
if command -v exa >/dev/null; then
  alias ls="exa"
  alias ll="exa -l --git"
  alias tree="exa --tree"
  alias gtree="exa -l --tree --git"
else
  alias tree='tree -a -I .git'
  alias ls="${aliases[ls]:-ls} -A"
fi

# Replace cat with bat
if command -v bat >/dev/null; then
  alias cat="bat"
fi

# Set shell options: http://zsh.sourceforge.net/Doc/Release/Options.html.
# Reference card: http://lanisolutions.com/wp-content/uploads/2011/05/oh-my-zsh_refcard.pdf
setopt glob_dots     # no special treatment for file names with a leading dot
setopt no_auto_menu  # require an extra TAB press to open the completion menu
```
3) Bootstrap docker and z4h:
```sh
# Get latest ZSH
brew install zsh

# Change user shell to the one installed by Homebrew
sudo dscl . -create "/Users/$USER" UserShell /usr/local/bin/zsh

# Install docker desktop
brew install docker --cask

# Download extra completions to ~/.dotfiles/completion
mkdir ~/.dotfiles/completion
curl -fsSL https://raw.githubusercontent.com/docker/docker-ce/master/components/cli/contrib/completion/zsh/_docker
 > ~/.dotfiles/completion/_docker
curl -fsSL https://raw.githubusercontent.com/docker/compose/master/contrib/completion/zsh/_docker-compose > ~/.dotfiles/completion/_docker-compose

# Install Z4H
sh -c "$(curl -fsSL https://raw.githubusercontent.com/romkatv/zsh4humans/v5/install)"
```
4) Run `zsh`
5) Type:
```sh
docker <tab>
docker-compose <tab>
```

For more information about my setup you can take a look into my dotfiles for reference: https://github.com/pdrastil/dotfiles
Bootstrap order: Brewfile > /run/before/*.sh > stow /symlinks > /run/after/*.sh

**Comment by romkatv** [comment]
Thanks for providing the instructions to reproduce the issue. I confirm that completions for `docker` and `docker-compose` don't work if I follow them. This is caused by the following step:

```zsh
# Install Z4H
sh -c "$(curl -fsSL https://raw.githubusercontent.com/romkatv/zsh4humans/v5/install)"
```

This step overrides `~/.zshrc`, so any modifications you've made to it are gone. If I use the original `~/.zshrc` that you've provided, completions work fine. It's likely that `source <(enterprise-tool completion zsh)` works fine, too.

I see the following piece of code in your dotfiles:

```zsh
if [ ! -d "$HOME/.cache/zsh4humans" ]; then
  subtitle "Installing Zsh for humans..."
  sh -c "$(download https://raw.githubusercontent.com/romkatv/zsh4humans/v5/install)"
fi
```

This is incorrect. You should run z4h installer only once in your life. The only thing it does is generate these 3 files:

- `~/.zshenv`
- `~/.zshrc`
- `~/.p10k.zsh`

Store these files in your dotfiles repository. In order to recreate your configuration on another machine, simply copy these files over. Do not run z4h installer unless you want to override `~/.zshenv` and `~/.zshrc`.

**Comment by pdrastil** [comment]
> This is incorrect. You should run z4h installer only once in your life. The only thing it does is generate these 3 files:

Thanks for pointing out the rewrite of `~/.zshrc` file during my bootstrap, however the installer is executed only once because installation directory is guarded by `$HOME/.cache/zsh4humans` condition and this directory doesn't exists on a new machine. If I hit my setup script again to update, installation will be skipped.

> This step overrides ~/.zshrc, so any modifications you've made to it are gone. If I use the original ~/.zshrc that you've provided, completions work fine. It's likely that source <(enterprise-tool completion zsh) works fine, too.

Sorry for invalid repro step. I've added `~/.zshrc` after installation and my bootstrap replaces original file with the symlink to `~/.dotfiles/zsh/.zshrc` when executed again. Can you please try again with content of zshrc provided in step 2 after the installation is done? Docker completions don't work even with proper file.

**Comment by romkatv** [comment]
> Can you please try again with content of zshrc provided in step 2 after the installation is done? Docker completions don't work even with proper file.

I'll try again in a few months. If you want a faster turn-around, provide *much* smaller instructions.

**Comment by pdrastil** [comment]
Thank you, but it's no longer needed. I've found minor typo in my local working changes that broke the fpath and I can verify that this works as expected.

**Comment by romkatv** [comment]
Thanks for the update.

Does `source <(enterprise-tool completion zsh)` work, too?

**Comment by pdrastil** [comment]
No sourcing completions from CLI tools is broken. Only workaround I was able to come with is to dump completions from the tool into file on fpath and then it works fine.

**Comment by romkatv** [comment]
> No sourcing completions from CLI tools is broken.

I don't think that's true. It may be broken for `enterprise-tool` but I cannot know why because I don't know what `enterprise-tool completion zsh` prints.

**Comment by romkatv** [comment]
Closing as fixed.
